compiler/rustc_middle/src/ty/consts.rs

```diff
@@ -74,6 +74,11 @@ impl<'tcx> Const<'tcx> {
 
         let ty = tcx.type_of(def.def_id_for_type_of());
 
+        let param_env = tcx.param_env(def.did);
```

Comment:
This normalization is causing a lot of cycle error. If I nest it into a `if TypeVisitable::has_projections(&ty) {`, it reduces greatly the number of such errors but it's not a good solution imo.

Do you have any idea what's going on here? Why would normalization would create cycle errors? (in code like `src/test/ui/variance/variance-associated-consts.rs`)

Reply:
ah yeah ofc that causes issues :sweat_smile: 

having a const parameter in the where clauses now relies on the `param_env` query to get the `ty::Const` but we need the `ty::Const` as part of the output of the `param_env` query.

Reply:
not really :sweat_smile: even changing the type of type level constants to not inherit their parents generics won't work because normalization depends on impl headers which can also contain `ty::Const`. Normalizing when creating `ty::Const` is too early :thinking: 

If we were to delay normalization of the type of consts to a later point this may mostly work. tbh that makes me think that we shouldn't land this change and should keep forbidding projections in the type of type system constants. Sorry for not noticing this issue earlier :sweat_smile: especially as we've had to deal with pretty much the same issue for `generic_const_exprs` as well.